primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,468 @@
+		let snooze = core::cmp::min(remainder.millis(), STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE);
+		sp_io::offchain::sleep_until(self.timestamp.add(Duration::from_millis(snooze)));
+	}
+}
+
+/// Storage based lock.
+///
+/// A lock that is persisted in the DB and provides a mutex behavior
+/// with a defined safety expirey deadline based on a [`Lockable`](Self::Lockable)
+/// implementation.
+pub struct StorageLock<'a, L>
+where
+	L: Lockable,
+{
+	// A storage value ref which defines the DB entry representing the lock.
+	value_ref: StorageValueRef<'a>,
+	deadline: L,
+}
+
+impl<'a, L> StorageLock<'a, L>
+where
+	L: Lockable,
+{
+	/// Create a new storage lock with [default expiry duration](Self::STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS).
+	pub fn new(key: &'a [u8]) -> Self {
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			deadline: L::deadline(),
+		}
+	}
+
+	fn try_lock_inner(&mut self, new_deadline: L) -> Result<(), Option<L>> {
+		let now = L::current();
+		let res = self
+			.value_ref
+			.mutate(|s: Option<Option<L>>| -> Result<L, Option<L>> {
+				match s {
+					// no lock set, we can safely acquire it
+					None => Ok(new_deadline),
+					// lock is set, but it's old. We can re-acquire it.
+					Some(Some(deadline)) if now.expired(&deadline) => Ok(new_deadline),
+					// lock is present and is still active
+					Some(Some(deadline)) => Err(Some(deadline)),
+					_ => Err(None),
+				}
+			});
+		match res {
+			Ok(Ok(_)) => Ok(()),
+			Ok(Err(_deadline)) => Err(None),
+			Err(e) => Err(e),
+		}
+	}
+
+	/// Attempt to lock the storage entry.
+	///
+	/// Returns a lock guard on success, otherwise an error containing `None` in
+	/// case the mutex was already unlocked before, or if the lock is still held
+	/// by another process `Err(Some(expiration_timestamp))`.
+	pub fn try_lock<'b>(&'b mut self) -> Result<StorageLockGuard<'a, 'b, L>, Option<L>>
+	where
+		'a: 'b,
+	{
+		match self.try_lock_inner(self.deadline) {
+			Ok(_) => Ok(StorageLockGuard::<'a, 'b> { lock: Some(self) }),
+			Err(e) => Err(e),
+		}
+	}
+
+	/// Try grabbing the lock until its expiry is reached.
+	///
+	/// Returns an error if the lock expired before it could be caught.
+	pub fn spin_lock<'b>(&'b mut self) -> StorageLockGuard<'a, 'b, L>
+	where
+		'a: 'b,
+	{
+		loop {
+			// blind attempt on locking
+			let deadline = match self.try_lock_inner(self.deadline) {
+				Ok(_) => return StorageLockGuard::<'a, 'b, L> { lock: Some(self) },
+				Err(Some(other_locks_deadline)) => other_locks_deadline,
+				_ => L::deadline(), // use the default
+			};
+			L::current().snooze(&deadline);
+		}
+	}
+
+	/// Explicitly unlock the lock.
+	pub fn unlock(&mut self) {
+		self.value_ref.clear();
+	}
+}
+
+/// RAII style guard for a lock.
+pub struct StorageLockGuard<'a, 'b, L: Lockable> {
+	lock: Option<&'b mut StorageLock<'a, L>>,
+}
+
+impl<'a, 'b, L: Lockable> StorageLockGuard<'a, 'b, L> {
+	/// Consume the guard but DO NOT unlock the underlying lock.
```

Comment:
Can you add docs describing the use case? I think I've elaborated on this in some of my previous comments.